File: resource/js/poker.js
function poker_check_valid_color() {
    var card_color_flop1 = document.getElementById('card_color_flop1');
    var card_color_flop2 = document.getElementById('card_color_flop2');
    var card_color_flop3 = document.getElementById('card_color_flop3');
    set_progress_rate(0, 100);
    if (card_color_flop1.value == 'none') {
        $("#card_color_flop1").css("border", "red 1px solid");
        return false;
    } else {
        $("#card_color_flop1").css("border", "")
        return true;
    }
    if (card_color_flop2.value == 'none') {
        $("#card_color_flop2").css("border", "red 1px solid");
        return false;
    } else {
        $("#card_color_flop2").css("border", "")
        return true;
    }
    if (card_color_flop3.value == 'none') {
        $("#card_color_flop3").css("border", "red 1px solid");
        return false;
    } else {
        $("#card_color_flop3").css("border", "")
        return true;
    }
}

function set_progress_rate(n, total) {
    //设置进度
    rate = (n / total * 100).toFixed(2)
    if (n >= 0) {
        $(".progress-bar").attr("aria-valuenow", n);
        $(".progress-bar").attr("aria-valuemax", total);
        $(".progress-bar").text(rate + "%");
        $(".progress-bar").css("width", rate + "%");
    }
}

function analyze_card() {
    set_progress_rate(0, 100);
    var canclick = true
    $('#poker_button').attr('disabled', canclick);
    var my_card1 = $.trim($("#card_num_my_hand1").val()) + $.trim($("#card_color_my_hand1").val());
    var my_card2 = $.trim($("#card_num_my_hand2").val()) + $.trim($("#card_color_my_hand2").val());
    var flop1 = $.trim($("#card_num_flop1").val()) + $.trim($("#card_color_flop1").val());
    var flop2 = $.trim($("#card_num_flop2").val()) + $.trim($("#card_color_flop2").val());
    var flop3 = $.trim($("#card_num_flop3").val()) + $.trim($("#card_color_flop3").val());
    var flop4 = $.trim($("#card_num_flop4").val()) + $.trim($("#card_color_flop4").val());
    var flop5 = $.trim($("#card_num_flop5").val()) + $.trim($("#card_color_flop5").val());

    console.log(my_card1, my_card2, flop1, flop2, flop3, flop4, flop5)

    var param = 'my_card1=' + my_card1
        + '&my_card2=' + my_card2
        + '&flop1=' + flop1
        + '&flop2=' + flop2
        + '&flop3=' + flop3
        + '&flop4=' + flop4
        + '&flop5=' + flop5;

    // var ajax = setInterval(function () {
    //     //每1秒请求一次进度
    //     $.ajax({
    //             url: "http://127.0.0.1:5001/progress_bar",
    //             type: "GET",
    //             cache: false,
    //             success: function (response) {
    //                 console.log(response);
    //                 n = response["n"];
    //                 var total = response["total"];
    //                 set_progress_rate(n, total);
    //
    //             }
    //         }
    //     );
    // }, 1000);

    $.post('/poker-run', param, function (data) {

            console.log(data)
            if (data == 'request form get error') {
                alert("解析参数失败")
            } else if (data == 'hand card empty') {
                alert("我的手牌中不能有空值，检查花色和牌面")
            } else if (data == 'flop card empty') {
                alert("转牌区必填3张表，转牌区的牌不可为空，检查花色和牌面")
            } else if (data == 'duplicated card find') {
                alert("有卡牌重复，请检查花色和牌面")
            } else {
                document.getElementById('overall_res').innerText = '我的胜率是 ' + data['my_win_rate'] + '，打平的概率是 ' + data['my_tie_rate'] + '，输掉的概率是 ' + data['my_lose_rate']
                document.getElementById('my_res').innerText = '牌型是【高牌】的概率是 ' + data['my_high_card_prob'] + '，牌型是【一对】的概率是 ' + data['my_pair_prob']
                    + '，牌型是【两对】的概率是 ' + data['my_two_pair_prob'] + '\n' + '牌型是【3条】的概率是 ' + data['my_three_kind_prob'] + '，牌型是【顺子】的概率是 ' + data['my_straight_prob']
                    + '，牌型是【同花】的概率是 ' + data['my_flush_prob'] + '\n' + '牌型是【葫芦】的概率是 ' + data['my_full_house_prob'] + '，牌型是【炸弹】的概率是 ' + data['my_four_kind_prob']
                    + '，牌型是【同花顺】的概率是 ' + data['my_straight_flush_prob'] + '\n' + '牌型是【皇家同花顺】的概率是 ' + data['my_royal_flush_prob']
                document.getElementById('enemy_res').innerText = '牌型是【高牌】的概率是 ' + data['enemy_high_card_prob'] + '，牌型是【一对】的概率是 ' + data['enemy_pair_prob']
                    + '，牌型是【两对】的概率是 ' + data['enemy_two_pair_prob'] + '\n' + '牌型是【3条】的概率是 ' + data['enemy_three_kind_prob'] + '，牌型是【顺子】的概率是 ' + data['enemy_straight_prob']
                    + '，牌型是【同花】的概率是 ' + data['enemy_flush_prob'] + '\n' + '牌型是【葫芦】的概率是 ' + data['enemy_full_house_prob'] + '，牌型是【炸弹】的概率是 ' + data['enemy_four_kind_prob']
                    + '，牌型是【同花顺】的概率是 ' + data['enemy_straight_flush_prob'] + '\n' + '牌型是【皇家同花顺】的概率是 ' + data['enemy_royal_flush_prob']
                document.getElementById('time_diff').innerText = data['time_diff'] + '秒'
            }
            document.getElementById("poker_button").disabled = false;
            set_progress_rate(100, 100);
            // clearInterval(ajax)
        }
    )
}

File: controller/poker.py
from flask import Blueprint, render_template, request, jsonify, make_response
from module.poker import *
import time
from multiprocessing import Pool


poker = Blueprint('poker', __name__)


@poker.route('/poker-run', methods=['POST'])
def poker_run_post():
    try:
        my_card1 = request.form.get('my_card1')
        my_card2 = request.form.get('my_card2')
        flop1 = request.form.get('flop1')
        flop2 = request.form.get('flop2')
        flop3 = request.form.get('flop3')
        flop4 = request.form.get('flop4')
        flop5 = request.form.get('flop5')
    except Exception as e:
        print(e)
        return 'request form get error'
    my_hand_list = [my_card1, my_card2]
    for i in my_hand_list:
        if 'none' in i:
            return 'hand card empty'
    for i in [flop1, flop2, flop3]:
        if 'none' in i:
            return 'flop card empty'
    card_list = []
    for i in [my_card1, my_card2, flop1, flop2, flop3, flop4, flop5]:
        if 'none' in i:
            pass
        else:
            card_list.append(i)
    if len(set(card_list)) < len(card_list):
        return 'duplicated card find'

    open_board_deck = []
    for i in [flop1, flop2, flop3, flop4, flop5]:
        if 'none' in i:
            pass
        else:
            open_board_deck.append(i)
    if len(set(card_list)) < len(card_list):
        return 'duplicated card find'
    try:
        print('调用成功')
        print(my_card1, my_card2, flop1, flop2, flop3, flop4, flop5)
        res = Poker().poker_run(my_hold_cards=[my_card1, my_card2], open_board_deck=open_board_deck,
                        enemy_hold_cards=[None, None])
        print(res)
        return res
    except Exception as e:
        print(e)
        return 'model excute failed'




# @poker.route('/progress_bar')
# def progress():
#     """查看进度"""
#     response = make_response(jsonify(dict(n=pbar.n, total=pbar.total)))
#     response.headers.add('Access-Control-Allow-Origin', '*')
#     response.headers.add('Access-Control-Allow-Headers', '*')
#     response.headers.add('Access-Control-Allow-Methods', '*')
#     return response
